File: 3/MaxPathSum.java
class MaxPathSum {
	public static int[] getMaxPathSumHelper(BinaryTree.Node root) {
		if(root == null) {
			return new int[]{0, 0};
		}
		
		int leftMax[] = getMaxPathSumHelper(root.left);
		int rightMax[] = getMaxPathSumHelper(root.right);
		int childBranchMax = Math.max(leftMax[1], rightMax[1]);
			
		int rootVal = root.data;

		int maxBranch = Math.max(childBranchMax + rootVal, rootVal);
		int maxSum = Math.max(leftMax[1] + rightMax[1] + rootVal, maxBranch);
		int maxPath = Math.max(Math.max(leftMax[0], rightMax[0]), maxSum);	

		return new int[]{maxPath, maxBranch};
	}
	
	public static int getMaxPathSum(BinaryTree.Node root) {
		int[] max = getMaxPathSumHelper(root); 
		return max[0];
	}

	public static void main(String args[]) {
		BinaryTree tree = new BinaryTree();

		for(int i=1; i<=7; i++) {
			tree.insert(i);
		}

		System.out.println(getMaxPathSum(tree.root));
	}
}

File: 3/js/src/MaxPathSum.ts
import {TreeNode, Tree} from "./Tree";

function getMaxPathSum(tree: Tree): number {
	return Math.max(...getMaxPathSumHelper(tree.root));
}

function getMaxPathSumHelper(root: TreeNode | null): [number, number] {
	if(root == null) {
		return [0,0];
	}

	const data = root.data;
	const [leftPathSum, leftBranchSum] = getMaxPathSumHelper(root.left);
	const [rightPathSum, rightBranchSum] = getMaxPathSumHelper(root.right);

	const maxChildBranch = Math.max(leftBranchSum, rightBranchSum);
	const maxBranchSum = Math.max(maxChildBranch + data, data);

	const maxSumWithRoot  = Math.max(data + leftBranchSum + rightBranchSum, maxBranchSum);

	const maxPathSum = Math.max(leftPathSum, rightPathSum, maxSumWithRoot);

	return [maxPathSum, maxBranchSum];
}

(function main() {
	const tree = new Tree();

	for(let i=1; i<=7; i++) {
		tree.insertOne(i);
	}

	console.log(getMaxPathSum(tree));
})();
